sdk/src/Core/Sampling/ServiceConnector.cs

```diff
@@ -119,47 +164,101 @@ public async Task<GetSamplingRulesResponse>  GetSamplingRules()
         /// <returns>Instance of <see cref="GetSamplingRulesResponse"/>.</returns>
         public async Task<GetSamplingTargetsResponse> GetSamplingTargets(List<SamplingRule> rules)
         {
-            GetSamplingTargetsRequest request = new GetSamplingTargetsRequest();
-            IList<Target> newTargets = new List<Target>();
             DateTime currentTime = TimeStamp.CurrentDateTime();
-            List<SamplingStatisticsDocument> samplingStatisticsDocuments = GetSamplingStatisticsDocuments(rules, currentTime);
-            request.SamplingStatisticsDocuments = samplingStatisticsDocuments;
-            Task<Model.GetSamplingTargetsResponse> responseTask;
+            List<SamplingStatisticsDocumentModel> samplingStatisticsDocumentModels = GetSamplingStatisticsDocuments(rules, currentTime);
+            var samplingStatisticsModel = new SamplingStatisticsModel();
+            samplingStatisticsModel.SamplingStatisticsDocuments = samplingStatisticsDocumentModels;
+
+            string requestContent = JsonMapper.ToJson(samplingStatisticsModel); // Marshall SamplingStatisticsDocument to json
+
+            Task<WebResponse> responseTask;
             lock (_xrayClientLock)
             {
                 RefreshEndPoint();
-                responseTask = _xrayClient.GetSamplingTargetsAsync(request);
+                responseTask = GetSamplingInfoAsync(_xrayConfig.ServiceURL + "/SamplingTargets", requestContent);
```

Comment:
Just getting the task here and doing the await outside of the lock defeats the purpose of the lock. The actual getting the response for the request will happen outside of the lock. So if the point of the lock was to not have multiple calls to the sevice url at the same time his won't work.

I assume this was done because you can't do an await with the lock keyword. Usually you get around that by directly using the .NET monitor class.

Reply:
Right, fixed in the new commit